File: server/Rydo.Api/Services/RideEventWindow.cs
using Rydo.Api.Data;

namespace Rydo.Api.Services;

/// <summary>
/// Shared 48-hour ride event window anchored on <see cref="Ride.ScheduledDate"/>.
/// Used by live ride, ride chat, and edit rules.
/// </summary>
public static class RideEventWindow
{
    public const int HoursAfterStart = 48;

    public static DateTime ClosesAt(Ride ride) =>
        ride.ScheduledDate.ToUniversalTime().AddHours(HoursAfterStart);

    public static bool HasStarted(Ride ride) =>
        DateTime.UtcNow >= ride.ScheduledDate.ToUniversalTime();

    /// <summary>Before start or within 48h after scheduled start.</summary>
    public static bool IsOpen(Ride ride) =>
        DateTime.UtcNow < ClosesAt(ride);

    public static bool ChatWritable(Ride ride) => IsOpen(ride);

    public static bool LiveAvailable(Ride ride) =>
        ride.Kind == RideKind.Scheduled && ride.RouteId != null && IsOpen(ride);

    public static bool CanEdit(Ride ride) => IsOpen(ride);

    public static bool CanEditScheduledDate(Ride ride) => !HasStarted(ride);

    public static object ToPayload(Ride ride) => new
    {
        closesAt = ClosesAt(ride).ToString("yyyy-MM-ddTHH:mm:ss.fffZ"),
        hasStarted = HasStarted(ride),
        liveAvailable = LiveAvailable(ride),
        chatReadOnly = !ChatWritable(ride),
        canEditScheduledDate = CanEditScheduledDate(ride),
    };
}

File: client/src/features/rides/utils/rideEventWindow.js
const WINDOW_MS = 48 * 60 * 60 * 1000;

/**
 * @param {{ scheduledDate?: string, time?: string, routeId?: number | null, rideKind?: string, rideEventWindow?: object } | null | undefined} ride
 */
export function rideEventWindow(ride) {
  if (!ride) {
    return {
      closesAt: null,
      hasStarted: false,
      isOpen: false,
      chatWritable: false,
      liveAvailable: false,
      canEditScheduledDate: false,
    };
  }

  if (ride.rideEventWindow && typeof ride.rideEventWindow === 'object') {
    const w = ride.rideEventWindow;
    const chatReadOnly = Boolean(w.chatReadOnly);
    return {
      closesAt: w.closesAt ?? null,
      hasStarted: Boolean(w.hasStarted),
      isOpen: !chatReadOnly,
      chatWritable: !chatReadOnly,
      liveAvailable: Boolean(w.liveAvailable),
      canEditScheduledDate: Boolean(w.canEditScheduledDate),
    };
  }

  const iso = ride.scheduledDate || ride.time;
  const start = iso ? new Date(iso) : null;
  if (!start || Number.isNaN(start.getTime())) {
    return {
      closesAt: null,
      hasStarted: false,
      isOpen: true,
      chatWritable: true,
      liveAvailable: Boolean(ride.routeId) && ride.rideKind !== 'soloLog',
      canEditScheduledDate: true,
    };
  }

  const closesAt = new Date(start.getTime() + WINDOW_MS);
  const now = Date.now();
  const hasStarted = now >= start.getTime();
  const isOpen = now < closesAt.getTime();
  const isScheduled = ride.rideKind !== 'soloLog';

  return {
    closesAt: closesAt.toISOString(),
    hasStarted,
    isOpen,
    chatWritable: isOpen && isScheduled,
    liveAvailable: isOpen && isScheduled && Boolean(ride.routeId),
    canEditScheduledDate: !hasStarted,
  };
}

/** @param {{ scheduledDate?: string, time?: string } | null | undefined} ride */
export function isRideUpcoming(ride) {
  const iso = ride?.scheduledDate || ride?.time;
  if (!iso) return true;
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return true;
  return d.getTime() >= Date.now();
}

/** Ride is in progress (started but still within the 48h window). */
export function isRideInProgress(ride) {
  const w = rideEventWindow(ride);
  return w.hasStarted && w.isOpen;
}
